### PR TITLE
Interactable configuration/documentation/test changes

### PR DESCRIPTION
**Pull request by Troy-Ferrell**
## Overview
This is a culmination of multiple PRs and merging the feature branch into MRTK_Dev:
- Related PRs: #5857 , #5889 , #5898 , #5925 , #5960 

There are a lot of fixes in here and I would reference each PR for details but the important notes for this PR is the review of breaking changes and validation steps.

## Upgraded Interactable and Visual Themes
Interactable has been refactored significantly. We took great care to not break anything, but it's a good idea before upgrading to 2.1 to back up your existing project. 

### Change Highlights:
* The Themes functionality with Interactable has been extracted into it's own feature called Visual Themes. 
* The Visual Themes code has been refactored to be more robust and simplified. This involves some class name changes and some property changes. 
* Themes can now target specific shader properties explicitly using the new `ThemeStateProperty.TargetShader` and `ThemeStateProperty.ShaderPropertyName` variables.
* Better documentation on how to create Interactables and Themes dynamically via code

**1) If you used default Theme scriptableobjects or created Theme assets using the default Theme types (like InteractableColorTheme):**

Everything should work as-is. To force your assets to update though, you can select your assets or scene/prefab items and view them in the inspector. Note: Unity only writes inspector changes to disk once the scene is saved. After doing this, you should be able to see the serialized diffs in your source control tracking.

**2) If you defined Events Receivers for your Interactables in scenes/prefabs:**

Everything should work as-is. To force your assets to update properly though, you can select and view the Events (foldout must be opened) to force the asset update. Note: Unity only writes inspector changes to disk once the scene is saved. 

**3) If you built a custom theme class extending InteractableThemeBase:**

The names of many classes have changed. See list below for full information. Three new properties should be implemented by your class extending `InteractableThemeBase`. 

`public override bool IsEasingSupported => true;`
Set to true if your theme supports easing between values when state changes

`public override bool AreShadersSupported => false;`
Set to true if your theme supports shader property targeting

`public override ThemeDefinition GetDefaultThemeDefinition()`
This method returns a default configuration object for your theme. Include properties, names, and types that your Theme is expecting to run when configured via `Init(GameObject host, ThemeDefinition definition)`

Look at \MixedRealityToolkit.SDK\Features\UX\Scripts\VisualThemes\ThemeEngines\ for examples of how to implement this method

Also function definitions may need to be updated for new class name which are arguments into the function such as
`InteractableThemeBase.Init(GameObject, ThemeDefinition)` 
`ThemePropertyValue GetProperty(ThemeStateProperty property)`
`SetValue(ThemeStateProperty property, int index, float percentage)`

**4) If you built a custom Event receiver class extending ReceiverBase:**

Everything should work as-is. Only step to be taken is to update any extended constructors to pass a name string into the base. Note: Fields must have the InspectorField attribute to show up on Interactable objects in the Unity inspector

**Change Details:**

- `InteractableThemeBase` now has abstract method `GetDefaultTheme()`, `IsEasingSupported`, `AreShadersSupported`
- Shader property metadata in Theme SO moved to `ThemeStateProperty.TargetShader` and `ThemeStateProperty.ShaderPropertyName` variables
- Event Receivers in scene/prefab files need to update to include AssemblyClassName. Bug where data was not originally saved.
- Multiple classes have been renamed see list below
- `InteractableTextureTheme` not extends `InteractableShaderTheme` and supports shader property targeting
- Elimination of `InteractableTypeFinder`/`InteractableTypesContainer` classes
- Creating Scriptable objects shortcut now under *Create* > *MRTK* > *Theme/State*

**Major API Renames:**
Interactable.Events => Interactable.InteractableEvents
Interactable.ForceUpdateThemes() => Interactable.RefreshSetup()
InteractableStateModel => BaseStateModel
InteractableThemePropertySettings => ThemeDefinition
InteractableThemeProperty => ThemeStateProperty
InteractableCustomSetting => ThemeProperty
InteractableThemePropertyValue => ThemePropertyValue
InteractableThemePropertyTypes => ThemePropertyType
States.CompareStates => States.Equals

## Changes
- Fixes: #5866 , #5652 , #5462 , #5058 , #5459 , #5050 , #5788 

## Verification
**Build following in GA and then update GA to feature branch**
1) Theme assets (SUCCESS)
2) CustomThemeClass (SUCCESS)
3) Interactable Events Receivers in a scene (SUCCESS)
4) Interactable Events Receivers a in prefab (SUCCESS)
5) Create custom EventReceiver class(SUCCESS)

**After MRTK code update, test the following**
Functionality works out-of-box after just hitting playing
Theme inspector works as expected
Interactable inspector works as expected

**Other validation**
All automated tests pass (SUCCESS)
Test InteractableExamples scene works (SUCCESS)
Test HandInteractionExample scene works (SUCCESS)
Tested import of Themes from RC2.1 (SUCCESS)

> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

### CODE REVIEW

File: Documentation/VisualThemes.md [hunk 1]
```diff
@@ -0,0 +1,281 @@
+
+# Visual Themes
+
+Themes allow for flexible control of UX assets in respond to various states transitions. This may involve changing a button's color, resizing an element in respond to focus, etc. The Visual Themes framework is made up of two key pieces: 1) configuration and 2) runtime engines.
```
Comment (ritijain) [inline Documentation/VisualThemes.md]: nit: "response"

File: Documentation/VisualThemes.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+To create a new [`State`](xref:Microsoft.MixedReality.Toolkit.UI.States) asset:
+
+1) Right click in the *Project Window*
+1) Select **Create** > **Mixed Reality Toolkit** > **State**
+
+![States ScriptableObject example in inspector](Images/Interactable/DefaultInteractableStates.png)
+
+A [`State`](xref:Microsoft.MixedReality.Toolkit.UI.States) ScriptableObject defines both the list of states as well as the type of *StateModel* to create for these states. A *StateModel* is a class that extends [`BaseStateModel`](xref:Microsoft.MixedReality.Toolkit.UI.BaseStateModel) and implements the state machine logic to generate the current state at runtime. The current state from this class is generally used by Theme Engines at runtime to dictate what values to set against material properties, GameObject transforms, and more.
```
Comment (ritijain) [inline Documentation/VisualThemes.md]: nit: can you also add a link to "State" documentation here  - pointing to the blurb in README_Interactable.md please.

File: Documentation/VisualThemes.md [hunk 3]
```diff
@@ -0,0 +1,281 @@
+
+# Visual Themes
```
Comment (ritijain) [inline Documentation/VisualThemes.md]: Love this documentation! It is really comprehensive

File: Documentation/VisualThemes.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+#### Mandatory implementations
+
+`public abstract void SetValue(ThemeStateProperty property, int index, float percentage)`
+(xref:Microsoft.MixedReality.Toolkit.UI.InteractableThemeBase.SetValue)
+
+For the given property, which can be identified by `ThemeStateProperty.Name`, set it's current state value on the targeted GameObject host (i.e set the material color, etc). The *index* indicates the current state value to access and the *percentage*, a float between 0 and 1, is used for easing/lerping between values.
+
+`public abstract ThemePropertyValue GetProperty(ThemeStateProperty property)`(xref:Microsoft.MixedReality.Toolkit.UI.InteractableThemeBase.GetProperty)
+
+For the given property, which can be identified by `ThemeStateProperty.Name`, return the current value set on the targeted Host  GameObject (i.e the current material color, the current local position offset, etc). This is primarily used for caching teh start value when easing between states.
```
Comment (ritijain) [inline Documentation/VisualThemes.md]: nit: typo - the

File: Documentation/README_Interactable.md [hunk 5]
```diff
@@ -2,416 +2,333 @@
...
+```
 
-<img src="../Documentation/Images/Interactable/StatesScriptableObject.png" width="450">
+## How to use Interactable
 
-The InteractableStates State Model will handle any state list with a layered ranking system, starting with the most isolated state and ending with the state that could contain all other states.
+### Building a simple button
 
-The DefaultInteractableStates list contains 4 states:
+One can create a simple button by simply adding the *Interactable* component to a GameObject that has a collider on it or on a child to receive input.
```
Comment (ritijain) [inline Documentation/README_Interactable.md]:
```suggestion
One can create a simple button by simply adding the *Interactable* component to a GameObject that has a collider on it or on a child to receive input.
```

Can you rephrase this to something like - "One can create a simple button by simply adding the *Interactable* component to a GameObject that is configured to receive input events. It can have a collider on it or on a child to receive input. If using interactable with a Unity UI based GameObjects it should be under the Canvas GameObject."

This is because i have seen a general practice of folks who prefer to use unity ui. The unity ui button does not provide as comprehensive a set of configuration options as the interactable and there is nothing really preventing Interactable from being used on unity ui components.

File: Documentation/VisualThemes.md (hunk 1 above)
Comment (keveleigh) [reply Documentation/VisualThemes.md]:
```suggestion
Themes allow for flexible control of UX assets in respond to various states transitions. This may involve changing a button's color, resizing an element in response to focus, etc. The Visual Themes framework is made up of two key pieces: 1) configuration and 2) runtime engines.
```

File: Documentation/VisualThemes.md (hunk 4 above)
Comment (keveleigh) [reply Documentation/VisualThemes.md]:
```suggestion
For the given property, which can be identified by `ThemeStateProperty.Name`, return the current value set on the targeted Host  GameObject (i.e the current material color, the current local position offset, etc). This is primarily used for caching the start value when easing between states.
```

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/ThemeEngines/InteractableColorTheme.cs [hunk 6]
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
```
Comment (ritijain) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/ThemeEngines/InteractableColorTheme.cs]: I am a little thrown off by how the InteractableColorTheme is setup. The fact that it inherits from InteractableShaderTheme is only helpful when the target is a Renderer. For other cases where the theme is attempting to change the text color, it does not leverage the InteractableShaderTheme. Can we split this into 2 classes - InteractableColorTheme - addressing Renderers, and InteractableFontColorTheme  - addressing teh text properties? 
(You can choose to Won't Fix it for this PR, but i would really like this to get addressed at some point)

File: Documentation/README_Interactable.md (hunk 5 above)
Comment (Troy-Ferrell) [reply Documentation/README_Interactable.md]: Good catch, I followed existing documentation content a little too closely here

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/ThemeEngines/InteractableColorTheme.cs (hunk 6 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/ThemeEngines/InteractableColorTheme.cs]: yea....you and me both. 

The problem is InteractableColorTheme is one of the more common themes. I mean the addition of the text types targeting was added based on user feedback of people wanting to use this theme. 

The problem with creating the new InteractableFontColorTheme is then existing users will need to migrate because their existing set-ups won't work anymore. And that's not something I can necessarily know ahead of time.

This is something I left more as-is because I'm already rocking the boat quite a lot with this refactor and this just seemed lower priority

I created #5971 to track this issue because yes I won't fix for now. I don't want to add more risk for this refactor that has already taken quite a lot of time

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Interactable.cs [hunk 7]
```diff
@@ -1102,12 +1127,15 @@ protected void IncreaseDimensionIndex()
 
             if (currentIndex != dimensionIndex)
             {
-                FilterThemesByDimensions();
+                SetupThemes();
                 forceUpdate = true;
             }
         }
 
-        public void ForceUpdateThemes()
```
Comment (ritijain) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Interactable.cs]: For a less disruptive change, instead of removing this method, would it make sense to mark it as Obsolete and redirect this to the new method?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/States/InteractableStates.cs [hunk 8]
```diff
@@ -108,7 +108,7 @@ public enum InteractableStateEnum
         public InteractableStates()
         {
             base.allStates = allStates;
-            currentState = allStates[0];
+            SetSate(allStates[0]);
```
Comment (ritijain) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/States/InteractableStates.cs]: nit: typo - rename to SetState

**Comment by ritijain** [comment]
For all the rename related breaking changes you have mentioned above, you could potentiallt do something like..

[Obsolete("OldClassName is obsolete. You should be using NewClassName instead")
public class OldClassName : NewClassName {.. ..}

to make them less disruptive.

**Comment by Troy-Ferrell** [comment]
> For all the rename related breaking changes you have mentioned above, you could potentiallt do something like..
> [Obsolete("OldClassName is obsolete. You should be using NewClassName instead")
> public class OldClassName : NewClassName {.. ..}
> to make them less disruptive.

Yea I considered that for the classes. I opted not doing that because 
1) I just want to cut clean now and not worry about someone coming along later to clean up these deprecated classes
2) The vast, vast majority of consumers atm with MRTK (which is still crazy small) should be ignorant of these classes that are sort of behind the scenes. it's not like renaming Interactable, ManipulationHandler, boundingbox, etc
3) Those who are leveraging these should be knowledgeable enough to make the necessary renames . It took me forever to learn the system. If someone else has actually done the same, they know the system well enough to understand what is going on. It's actually nice the compiler will point out where that is necessary to rename. Furthermore, this small group of advanced developers are unlikely to have a large amount of custom theme engines. It'll likely be one or two classes. 
4) I thought about leaving the old class names as empty classes in one cs file but then you would still get compiler errors for the missing fields. Again, I would rather just cut clean now.

I wish I could rename all the InterctableActivateTheme, InteractableColorTheme but that would be a harder break and annoying to migrate

Thanks for the review! Very helpful feedback

**Comment by Troy-Ferrell** [comment]
Performed migration from GA to feature branch code using the attached assets to test:
[MigrationTest.zip](https://github.com/microsoft/MixedRealityToolkit-Unity/files/3628533/MigrationTest.zip)

**Comment by julenka** [comment]
I've been testing this out, I was wondering, why does interactable have a States field, but then also have a separate States field for each theme? It sounds like it's possible for themes to have a state machine different from interactable. How is the States field in Interactable itself used? Is it used for event receivers?

image

**Comment by julenka** [comment]
I hit a bug when modifying Interactable. To repro:

1. Use the following theme:

image

2. Hit "Add Theme Definition" Twice
3. Try to delete one of the theme definitions by pressing "Delete" button

I get the following call stack:

ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
System.ThrowHelper.ThrowArgumentOutOfRangeException (System.ExceptionArgument argument, System.ExceptionResource resource) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.ThrowHelper.ThrowArgumentOutOfRangeException () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Collections.Generic.List`1[T].RemoveAt (System.Int32 index) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.ClearHistoryCache (System.Int32 index) (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:501)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.RenderDeleteButton (System.Int32 index) (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:430)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.RenderThemeDefinitions () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:121)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.RenderTheme () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:70)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.OnInspectorGUI () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:53)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector.RenderProfileSettings () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:173)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector.RenderCustomInspector () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:94)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector.OnInspectorGUI () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:83)
UnityEditor.InspectorWindow.DoOnInspectorGUI (System.Boolean rebuildOptimizedGUIBlock, UnityEditor.Editor editor, System.Boolean wasVisible, UnityEngine.Rect& contentRect) (at C:/buildslave/unity/build/Editor/Mono/Inspector/InspectorWindow.cs:1625)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)

**Comment by julenka** [comment]
Found a bug when adding a new theme via "create and assign theme" button. Got the following stack:

NullReferenceException: Object reference not set to an instance of an object
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.AddThemeDefinition () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:450)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.OnEnable () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:42)
UnityEditor.Editor:CreateEditor(Object)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector:RenderProfileSettings() (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:172)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector:RenderCustomInspector() (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:94)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector:OnInspectorGUI() (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:83)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)

To repro: 
1. Add second profile
2. Try to create a new theme definition

**Comment by killerantz** [comment]
The dream for states would be defining your own, then creating a script to determine state based on input into Interactable or from other sources. The easiest way to get the system working was to create an enum and then drive some of the state setting in Interactable, while exposing ways to set some states externally.

Themes and receivers can see what states are being set by Intractable and what the current state is and then react. State is more tied to Interactable now then it is to themes. In themes we care about the name of each state so the designer understands what each value relates to, the values are also stored for each state in the theme, but how they are set or what the represent are abstracted to anything that derives InteractableThemeBase.

The individual state meanings make more sense for receivers though, so that was the main reason for the enum and setting state in Interactable.

**That said, it would be interesting to move states out of Interactable and directly on to themes.** Interactable sets hasFocus, hasPress and other properties it cares about, but I never wanted the theme or state systems to be restricted to just Interactable states, that's why there are several that states (Collision) have nothing to do with Interactable. I also didn't want themes to be limited to only the states I could think of today.

Moving states to just the theme system could be the way to go. The state script could look at Interactable from the theme and set it's own state, it could also look at other classes on the component and set any new states a designer comes up with. This would would make the state system more flexible, and remove the need for the enum. The down side would be that a programmer will be needed to setup any code required for newly created states (which would happen regardless) and we would need to have a clear path for someone to set state externally. Not just through code, but the inspector as well.

Receivers could just look at properties in Interactable or properties on other classes and achieve the same result they are today.

**Comment by Troy-Ferrell** [comment]
> I've been testing this out, I was wondering, why does interactable have a States field, but then also have a separate States field for each theme? It sounds like it's possible for themes to have a state machine different from interactable. How is the States field in Interactable itself used? Is it used for event receivers?
> 
> image

So this is part of the original architecture except both States properties were not surfaced. 
**Theme** => Uses it's states property to determine how many values a property needs to define (for example, what color to set for each state)
**Interactable** => Uses States more to determine what kind of State Machine (i.e state model class) to instantiate at runtime inside it's class. Then it updates that state machine instance. 

I throw a warning that the States properties need to be the same. Technically they don't because the values are accessed just based on index. So if I had a states object with just Default/Focus, it should theoretically work against an object with Default/Focus/Pressed/Disabled. 

States in general requires a lot of refactoring. This is just something I didn't want to untangle and add risk/work to this PR. 

It's weird but short term I don't know how to fix it. I can't just enforce only "DefaultInteractableStates" because we actually use the "FocusPressTouchGrabStates" as well as Interactable.

**Comment by Troy-Ferrell** [comment]
@killerantz , the Interactable though requires States because it creates the State Model. And that information is largely driven to Receivers/Themes. 

The problem is there still needs to be some interface/agreement mechanism between all parties. Interactable/recievers can't just use themes because you can have multiple Themes which may have different overlaping states.

**Comment by Troy-Ferrell** [comment]
> I hit a bug when modifying Interactable. To repro:
> 
> 1. Use the following theme:
> 
> image
> 
> 1. Hit "Add Theme Definition" Twice
> 2. Try to delete one of the theme definitions by pressing "Delete" button
> 

I cannot repo this as of yet.

**Comment by Troy-Ferrell** [comment]
> Found a bug when adding a new theme via "create and assign theme" button. Got the following stack:
> To repro:
> 
> 1. Add second profile
> 2. Make the target object a TextMesh object that's not a descendant of the cube where interactable is located.
> 3. Try to create a new theme definition

This has been fixed

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/States/States.cs [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Serialization;
+
+namespace Microsoft.MixedReality.Toolkit.UI
+{
+    [CreateAssetMenu(fileName = "States", menuName = "Mixed Reality Toolkit/State", order = 1)]
+    public class States : ScriptableObject
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/States/States.cs]: Please add doccomment

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/States/States.cs [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        private Type stateModelType;
+
+        [FormerlySerializedAs("StateLogicName")]
+        [SerializeField]
+        private string StateModelClassName;
+
+        [SerializeField]
+        private string AssemblyQualifiedName;
+
+        public States()
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/States/States.cs]: Please make all public methods have doccomments

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/ThemeEngines/InteractableActivateTheme.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI
+{
+    public class InteractableActivateTheme : InteractableThemeBase
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/ThemeEngines/InteractableActivateTheme.cs]: Please add doccomments for all public fiends.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Events/InteractableEvent.cs [hunk 12]
```diff
@@ -15,98 +17,75 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
+        [NonSerialized]
+        private Type receiverType;
 
-            return data;
-        }
+        [SerializeField]
+        private string ClassName;
+
+        [SerializeField]
+        private string AssemblyQualifiedName;
```
[the remark this answers is not in the document]
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Events/InteractableEvent.cs]: yes this is the main thing in event receivers. This was not correctly being set in GA so it should be blank in lots of people's code. That is why I have the new editor/runtime code paths update appropriately and the backcode in the EventType

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Interactable.cs [hunk 13]
```diff
@@ -46,28 +45,39 @@ protected static IMixedRealityInputSystem InputSystem
...
+        protected readonly HashSet<IMixedRealityInputSource> pressingInputSources = new HashSet<IMixedRealityInputSource>();
+
         /// <summary>
         /// Is the interactable enabled?
         /// </summary>
         public bool Enabled = true;
 
+        [FormerlySerializedAs("States")]
+        [SerializeField]
+        private States states;
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Interactable.cs]: I've been testing this out, I was wondering, why does interactable have a States field, but then also have a separate States field for each theme? It sounds like it's possible for themes to have a state machine different from interactable. How is the States field in Interactable itself used? Is it used for event receivers?

image

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs [hunk 14]
```diff
@@ -0,0 +1,613 @@
...
+            EditorUtility.SetDirty(theme);
+        }
+
+        #endregion
+
+        #region Theme Definition History
+
+        protected void ClearHistoryCache(int index)
+        {
+            if (theme == null || theme.History == null || index >= theme.History.Count)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs]: I hit a bug when modifying Interactable. To repro:

Use the following theme:
image


Hit "Add Theme Definition" Twice
Try to delete one of the theme definitions by pressing "Delete" button
I get the following call stack:

ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
System.ThrowHelper.ThrowArgumentOutOfRangeException (System.ExceptionArgument argument, System.ExceptionResource resource) (at :0)
System.ThrowHelper.ThrowArgumentOutOfRangeException () (at :0)
System.Collections.Generic.List`1[T].RemoveAt (System.Int32 index) (at :0)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.ClearHistoryCache (System.Int32 index) (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:501)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.RenderDeleteButton (System.Int32 index) (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:430)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.RenderThemeDefinitions () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:121)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.RenderTheme () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:70)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.OnInspectorGUI () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:53)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector.RenderProfileSettings () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:173)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector.RenderCustomInspector () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:94)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector.OnInspectorGUI () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:83)
UnityEditor.InspectorWindow.DoOnInspectorGUI (System.Boolean rebuildOptimizedGUIBlock, UnityEditor.Editor editor, System.Boolean wasVisible, UnityEngine.Rect& contentRect) (at C:/buildslave/unity/build/Editor/Mono/Inspector/InspectorWindow.cs:1625)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)

Deleting the second to last theme definition worked, but deleting the very last theme definition did not work, probably an easy fix then.

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs [hunk 15]
```diff
@@ -0,0 +1,613 @@
...
+            return false;
+        }
+
+        #endregion
+
+        #region Theme Definition Management
+
+        protected virtual void AddThemeDefinition()
+        {
+            Type defaultType = typeof(InteractableActivateTheme);
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs]: Found a bug when adding a new theme via "create and assign theme" button. Got the following stack:

NullReferenceException: Object reference not set to an instance of an object
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.AddThemeDefinition () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:450)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.OnEnable () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:42)
UnityEditor.Editor:CreateEditor(Object)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector:RenderProfileSettings() (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:172)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector:RenderCustomInspector() (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:94)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector:OnInspectorGUI() (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:83)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)

To repro:

Add second profile
Make the target object a TextMesh object that's not a descendant of the cube where interactable is located.
Try to create a new theme definition

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Interactable.cs (hunk 13 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Interactable.cs]: So this is part of the original architecture except both States properties were not surfaced.
Theme => Uses it's states property to determine how many values a property needs to define (for example, what color to set for each state)
Interactable => Uses States more to determine what kind of State Machine (i.e state model class) to instantiate at runtime inside it's class. Then it updates that state machine instance.

I throw a warning that the States properties need to be the same. Technically they don't because the values are accessed just based on index. So if I had a states object with just Default/Focus, it should theoretically work against an object with Default/Focus/Pressed/Disabled.

States in general requires a lot of refactoring. This is just something I didn't want to untangle and add risk/work to this PR.

It's weird but short term I don't know how to fix it. I can't just enforce only "DefaultInteractableStates" because we actually use the "FocusPressTouchGrabStates" as well as Interactable.

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs (hunk 14 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs]: This has been fixed. Issue with the if condition not checking >=

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs (hunk 15 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs]: This has been fixed

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Interactable.cs (hunk 13 above)
Comment (killerantz) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Interactable.cs]: The theme should hide the State field when rendered in the Interactable UI. It's part of Themes for the reason you mentioned above, to allow the theme to have something to configure without being placed on an Interactable.

From what I remember, there should be an issue with placing a theme with one States object on an Interactable with another States object containing different states. Or have a single theme applied to two Interactables with different state objects. Haven't tested that though.

Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Interactable.cs]: I changed it to not hide. I want it to be obvious to the user. Also, I am trying to veer away from custom render logic. It should just be here is a Theme and render like it you normally would. If the States objects are different, it renders a warning message. 

Furthermore, if I hide it, the user has to double click into the theme to change it anyways which is just more work.

Comment (killerantz) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Interactable.cs]: Good point. 

"If the States objects are different, it renders a warning message." this is cool.

File: Documentation/README_Interactable.md [hunk 16]
```diff
@@ -2,416 +2,328 @@
...
 {
-     renderer.GetPropertyBlock(propertyBlock);
-
-     int propId = property.GetShaderPropertyId();
-
-     Color newColor = Color.Lerp(property.StartValue.Color, property.Values[index].Color, percentage);
-     block.SetColor(propId, color);
+    new InteractableProfileItem()
+    {
+        Themes = new List<Theme>() { <insert Theme asset here> },
```
Comment (julenka) [inline Documentation/README_Interactable.md]: Can you actually show how to load / instantiate a theme asset as well? Without that I'm having difficulty creating an interactable at runtime.

**Comment by Troy-Ferrell** [comment]
@julenka, fixed that and updated docs! Added Interactable.GetDefaultThemeAsset()
